src/camera.c

```diff
@@ -19,11 +19,205 @@ CameraControlSettings* test_ray_zone(f32 posX, f32 posY, f32 posZ, Collider** zo
     }
 }
 
-s32 func_800322DC(f32, f32, f32, f32, f32, f32, f32, f32, f32*, f32*, f32*);
-INCLUDE_ASM(s32, "camera", func_800322DC);
+s32 calculate_segment_intersection(f32 x1, f32 z1, f32 x2, f32 z2, f32 x3, f32 z3, f32 x4, f32 z4, f32* interX, f32* interZ, f32* squared_dist) {
+    f32 dx14;
+    f32 dx13;
+    f32 dz;
+    f32 dz12;
+    f32 dx34;
+    f32 dx12;
+    f32 dx;
+    f32 alpha;
+    f32 minus_dz12;
+    f32 p4_side;
+    f32 x;
+    f32 p3_side;
+    f32 z;
+    f32 dx12_copy;
+    f32 dz13;
+    f32 dz14;
+    f32 dz34;
+    f32 p3_side_;
+
+    dx12_copy = x2 - x1;
+    dx12 = dx12_copy;
+    dz12 = z2 - z1;
+    dx13 = x3 - x1;
+    dz13 = z3 - z1;
+    dx14 = x4 - x1;
+    dz14 = z4 - z1;
+    minus_dz12 = -dz12;
+    dx34 = x4 - x3;
+    dz34 = z4 - z3;
+
+    // if distance between points on the line is 0
+    if (dx12_copy == 0.0f && dz12 == 0.0f) {
+        return FALSE;
+    }
+    // if length of second segment is 0
+    if (dx34 == 0.0f && dz34 == 0.0f) {
+        return FALSE;
+    }
+
+    p3_side_ = minus_dz12 * dx13 + dx12_copy * dz13;
+    if (p3_side_ < 0.0f) {
+        p3_side = -1.0f;
+    } else if (p3_side_ > 0.0f) {
+        p3_side = 1.0f;
+    } else {
+        p3_side = 0.0f;
+    }
+    p3_side_ = p3_side;
+
+    dx = minus_dz12 * dx14 + dx12_copy * dz14;
+    if (dx < 0.0f) {
+        p4_side = -1.0f;
+    } else if (dx > 0.0f) {
+        p4_side = 1.0f;
+    } else {
+        p4_side = 0.0f;
+    }
+
+    // P3 and P4 are on the same side relative to the line: no intersection
+    if (p3_side_ == p4_side) {
+        return FALSE;
+    }
+
+    if (fabsf(dx12) > fabsf(dx34)) {
+        /*
+        We represent intersection point P as P = P3 + alpha * (P4 - P3)
+        and solve the equation (P - P1) x (P2 - P1) = 0, where 'x' is cross product, P1 and P2 are points on the line
+        and P3 and P4 are the ends of the segment.
+        So, (P3 - P1 + alpha * (P4 - P3)) x (P2 - P1) = 0,
+        alpha = [-(P3 - P1) x (P2 - P1)] / [(P4 - P3) x (P2 - P1)]
+        */
+        // same as -(P3 - P1) x (P2 - P1)
+        alpha = x1 * dz12 - z1 * dx12 + dx12 * z3 - dz12 * x3;
```

Comment:
I think it would be better to use the same notation for the variables as you do for the comments.
i.e., P1x, P1z instead of x1, z1.
Further, consider calling the points A1, A2, B1, B2, instead of P1/2/3/4 to indicate which line segment each belongs to.